OC-4907 Clients and Users should accept public_key

hosh opened the pull request:
- create users should accept public_key
- create clients should accept public_key

No changes to chef_wm were required to get update users and update clients to accept public_key

See:
- https://github.com/opscode/chef_wm/pull/39
- https://github.com/opscode/chef_objects/pull/26
- https://github.com/opscode/chef-pedant/pull/8

hosh committed:
POST /users should accept public_key

Generate keypair only if both "certificate" and "public_key"
fields in the json are undefined.

hosh committed:
POST /clients should accept public_key

Generate keypair only when certificate and public_key
fields are undefined in the json

seth commented:
Looks good.

In the case of creating a user or a client with a client-side generated key pair, the returned response will have a `private_key` field with a value of null. I can see that doing so makes for a minimal change. I think it would be slightly cleaner to only return the private_key field when we have a key to provide.

Thoughts? Perhaps a small helper to build the EJSON?

hosh commented:
Ok, I'll see what I can do. Not putting in private_key on the return makes sense, less confusing for people writing code against the raw API.

seth commented:
Sounds good. There are a few examples that might be relevant in chef_client:assemble_client_ejson among possibly other places.

hosh committed:
Use chef_object:set_key_pair() and set_public_key()

  - Updated endpoints to use the consolidated set_key_pair() and
    set_public_key() functions
  - Use undefined instead of null atom, so that set_key_pair() will skip
    setting the private key

hosh commented:
@seth since everything filtered through `set_key_pair()` (or can go through it), I figured it was easier to consolidate and modify that behavior rather than create something out of assemble_client

seth commented:
MWR
